File: src/leetcode/binary_search/Search_a_2D_Matrix_II_240.java
package leetcode.binary_search;

/**
 * Created by sunbo_000 on 10/14/2016.
 */

/*
    https://leetcode.com/problems/search-a-2d-matrix-ii/
 */
public class Search_a_2D_Matrix_II_240 {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) return false;

        int m = matrix[0].length - 1;
        int n = 0;
        while (n < matrix.length && m >= 0) {
            if (matrix[n][m] == target) return true;
            else if (matrix[n][m] > target) m--;
            else n++;
        }
        return false;

    }


    public static void main(String[] args) {
        Search_a_2D_Matrix_II_240 solution = new Search_a_2D_Matrix_II_240();
        int[][] matrix = new int[][]{
                {1, 3, 5, 7, 9},
                {2, 4, 6, 8, 10},
                {11, 13, 15, 17, 19},
                {12, 14, 16, 18, 20},
                {21, 22, 23, 24, 25}
        };
        /*int[][] matrix = new int[][]{
                {1, 1}
        };*/
        System.out.println(solution.searchMatrix(matrix, 8));
    }
}

File: src/leetcode_javascript/Search_a_2D_Matrix_II_240.js
/**
 * 240 Search a 2D Matrix II
 *
 * Write an efficient algorithm that searches for a value in an m x n matrix. This matrix has the following properties:

  Integers in each row are sorted in ascending from left to right.
  Integers in each column are sorted in ascending from top to bottom.
  For example,

  Consider the following matrix:

  [
    [1,   4,  7, 11, 15],
    [2,   5,  8, 12, 19],
    [3,   6,  9, 16, 22],
    [10, 13, 14, 17, 24],
    [18, 21, 23, 26, 30]
  ]
  Given target = 5, return true.

  Given target = 20, return false.
 *
 * https://leetcode.com/problems/search-a-2d-matrix-ii/
 */

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if (matrix === null || matrix.length === 0 || matrix[0].length === 0) {
        return false;
    }
    var m = 0;
    var n = matrix[0].length - 1;
    while (m < matrix.length && n >= 0) {
        if (matrix[m][n] === target) {
            return true;
        }
        if (matrix[m][n] > target) {
            n--;
        } else {
            m++;
        }
    }
    return false;
};